### DISCUSSION TITLE
Availability of CEF's new "chrome" browser layer

### DISCUSSION THREAD

**Discussion by Auersberg** [cefsharp/CefSharp#3469] [General]
Marshall Greenblatt, founder of the CEF project, is currently making some architectural changes:
https://bitbucket.org/chromiumembedded/cef/issues/2969/support-chrome-windows-with-cef-callbacks

Some of the changes already come with CEFSharp 90.1.0 (https://www.myget.org/gallery/cefsharp). For example, this package already includes the change from cef*.pak to chrome*.pak that Marshall announced:
https://bitbucket.org/chromiumembedded/cef/commits/117499421184

Since these changes would include some features I want to use for my developments (e.g. DataList), I wonder if it is already possible to test the changes with release 90.1.0, or if this is planned in the future. In a first test I could not detect any changes.

**Comment by amaitland** [comment]
It's my understanding this feature is considered experimental and not ready for production use.

>  if this is planned in the future

Yes, when this feature is ready it will be supported. 

 As for testing it now a quick search suggests you need to set CefSettings.chrome_runtime at runtime. This property isn't exposed in CefSharp yet, you'll have to modify/compile the project from source. 

https://github.com/chromiumembedded/cef/blob/f7a4c777e838d9a4f3f62d122ee642313c6888dd/include/internal/cef_types.h#L200

**Reply by Auersberg** [reply]
Thanks Alex, will give it a try.
Yes, I know it's not ready for production yet, but I'll have to keep an eye on these changes as they might be relevant to us. Btw, we will make a donation to the project to support your great work.

**Reply by Auersberg** [reply]
For sake of completeness for other people interested::

The command line argument "--enable-chrome-runtime" sets the flag for using the Chromium runtime. Code:
https://github.com/chromiumembedded/cef/blob/a6a8c0e845e8a2c85ef1f8ba0dd996fb39080ff0/tests/cefclient/browser/main_context_impl.cc#L96
and Marshall's note here:
https://bitbucket.org/chromiumembedded/cef/issues/2969/support-chrome-windows-with-cef-callbacks

I executed a test with this setting and could not see any changes, but I will keep an eye on it:
```
CefSettings settings = new CefSettings();
settings.CefCommandLineArgs.Add("enable-chrome-runtime");
```

**Reply by amaitland** [reply]
> Btw, we will make a donation to the project to support your great work.

@Auersberg That would be greatly appreciated, thank you!

There's been quite a bit of progress on https://bitbucket.org/chromiumembedded/cef/issues/2969/support-chrome-windows-with-cef-callbacks in the last week or so.

I've upgraded `master` to `90.4.2+gd4386da+chromium-90.0.4430.51 / Chromium 90.0.4430.51` and added the `CefSettings.ChromeRuntime` property in commit https://github.com/cefsharp/CefSharp/commit/94fb45766de8e0b809bc9ef9195c4e0a9259919f

> The command line argument "--enable-chrome-runtime" sets the flag for using the Chromium runtime. Code:
> https://github.com/chromiumembedded/cef/blob/a6a8c0e845e8a2c85ef1f8ba0dd996fb39080ff0/tests/cefclient/browser/main_context_impl.cc#L96

The `--enable-chrome-runtime` is only implemented in the `CEF Sample Applications(cefclient and cefsimple)`. 

You can enable the `Chrome Runtime` programmatically when using `CefSharp` (you'll need a `CI Build` available from https://github.com/cefsharp/CefSharp#releases-1 the `90.4.20-CI4007` build is based on `90.4.2+gd4386da+chromium-90.0.4430.51 / Chromium 90.0.4430.51`):

```c#
var settings = new CefSettings();
settings.ChromeRuntime = true;
Cef.Initialize(settings);
```

Using the `CefSharp.WinForms.Example` is crashing when `ChromeRuntime` is enabled, appears we're using features that aren't implemented or don't work properly, I'll see if I can get at least a basic example working when I have a minute. Will let you know how I go.

Is there a particular feature you were after?

**Comment by Auersberg** [comment]
Hi Alex, sorry for my late reply, but we had to develop our applications further. 
For us, only the DataList feature of the HTML text input field is currently interesting. We want to give the user the option to suggest possible inputs when filling out a text field. https://developer.mozilla.org/de/docs/Web/HTML/Element/datalist
This only works with the new Chrome runtime. However, I have already noticed in the test that the runtime does not work in our WinForms application. 

Apart from that, this project has helped us a lot and we will give you a donation at the end of our project.

**Reply by amaitland** [reply]
> However, I have already noticed in the test that the runtime does not work in our WinForms application.

The chrome runtime doesn't yet support parenting to a non-views windows (In this case embedding the browser in a WinForms control/form is not yet supported). Until CEF adds support it won't be possible to embed a chrome runtime browser in WinForms (or WPF).

> We want to give the user the option to suggest possible inputs when filling out a text field. https://developer.mozilla.org/de/docs/Web/HTML/Element/datalist

You should be able to use a JavaScript [polyfill](https://developer.mozilla.org/en-US/docs/Glossary/Polyfill) to add the missing functionality. https://github.com/thgreasi/datalist-polyfill looks fairly detailed, I haven't tested to see if this particular implementation is viable. 

> Apart from that, this project has helped us a lot and we will give you a donation at the end of our project.

Thank you, greatly appreciated!

**Reply by amaitland** [reply]
I'm hopeful that in `M102` we'll be able to use the `Chrome Runtime` with `WinForms` (won't be enabled by default). 

https://bitbucket.org/chromiumembedded/cef/issues/3294/chrome-add-support-for-embedded-non-views

**Reply by amaitland** [reply]
I've created #4123 which I'll use to compile implementation specific details about the `Chrome` runtime.
